src/components/tools/edit/formatters.js

```diff
@@ -0,0 +1,325 @@
+                );
+            }
+
+            if (container_volumes) {
+                values.container.container_volumes = container_volumes.map(
+                    ({ container_path, host_path }) => ({
+                        container_path,
+                        host_path,
+                    })
+                );
+            }
+
+            if (container_volumes_from) {
+                values.container.container_volumes_from = container_volumes_from.map(
+                    ({ name, name_prefix, tag, url, read_only }) => ({
+                        name,
+                        name_prefix,
+                        tag: tag || "",
+                        url: url || "",
+                        read_only,
+                    })
+                );
+            }
+        }
+    }
+
+    return values;
+}
+
+/**
+ * Formats the form values for submission to the service.
+ *
+ * @param {Tool} values
+ * @param {Object} config
+ * @param {Object} config.vice
+ * @param {string} config.vice.defaultCasUrl
+ * @param {string} config.vice.defaultName
+ * @param {string} config.vice.defaultImage
+ * @param {string} config.vice.defaultCasValidate
+ * @param {boolean} isAdmin
+ *
+ * @returns The Tool formatted for submission to the service.
+ */
+export function formatSubmission(values, config, isAdmin) {
+    const {
+        id,
+        name,
+        version,
+        attribution,
+        description,
+        type,
+        location,
+        restricted,
+        time_limit_seconds,
+        container,
+        implementation,
+    } = values;
+
+    const submission = {
+        id,
+        name,
+        description,
+        version,
+        type,
+        interactive: type === TOOL_TYPES.INTERACTIVE,
+    };
+
+    const {
+        name: containerName,
+        entrypoint,
+        working_directory,
+        uid,
+        min_cpu_cores,
+        min_memory_limit,
+        cpu_shares,
+        max_cpu_cores,
+        memory_limit,
+        min_disk_space,
+        pids_limit,
+        container_ports,
+        container_devices,
+        container_volumes,
+        container_volumes_from,
+    } = container;
+
+    // An undefined value means that key won't be submitted in the request,
+    // and container settings will be removed if its key is not included in the
+    // tool update request.
+    submission.container = {
+        ...container,
+        entrypoint: entrypoint || undefined,
+        working_directory: working_directory || undefined,
+        uid: uid || undefined,
+        max_cpu_cores: max_cpu_cores || undefined,
+        memory_limit: memory_limit || undefined,
+        min_disk_space: min_disk_space || undefined,
+        pids_limit: pids_limit || undefined,
+        skip_tmp_mount: false,
+        interactive_apps: undefined,
+        container_ports: undefined,
+        container_devices: undefined,
+        container_volumes: undefined,
+        container_volumes_from: undefined,
+    };
```

Comment:
The `PATCH` endpoint docs have the following note:
> If the container object is omitted in the request, then existing container settings will not be modified, but if the container object is present in the request, then all container settings must be included in it. Any existing settings not included in the request's container object will be removed.

So I'm using `undefined` to remove container settings in submissions (something in the client will remove those keys at some point in the serialization and submission to the service).

This only applies to container settings, so for top-level keys I'm still allowing empty strings (or `0`) to allow those values to be removed, which appears to be acceptable to the service.